<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [2]:
import pandas as pd
import psycopg2

In [ ]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

In [3]:

DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [40]:
query = f'''select * 
            from vacancies           
'''
df = pd.read_sql_query(query, connection)
df.sample(1)

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_1184\894878095.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


id                            name        key_skills  \
27921  55440563  Frontend developer (JS, React)  HTML5\tJS\treact   

            schedule          experience        employment  salary_from  \
27921  Гибкий график  От 1 года до 3 лет  Полная занятость          NaN   

       salary_to  area_id  employer_id  
27921        NaN        1       548557

In [52]:
query = f'''
            select distinct count(id)
            
            from vacancies           
'''
df = pd.read_sql_query(query, connection)
df.sample(1)

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_1184\4132213246.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0  49197

Количество вакансий 49 197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [65]:
query = f'''
            select distinct count(name)
            
            from EMPLOYERS
                   
'''
df1 = pd.read_sql_query(query, connection)
df1.sample(1)

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_1184\779171514.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df1 = pd.read_sql_query(query, connection)


count
0  23501

Количество работадетелей 23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [69]:
query = f'''
            select distinct count(name)
            
            from areas
                   
'''
df2 = pd.read_sql_query(query, connection)
df2.sample(1)

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_1184\3115053219.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df2 = pd.read_sql_query(query, connection)


count
0   1362

Количество регинов 1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [70]:
query = f'''
            select distinct count(name)
            
            from industries
                   
'''
df3 = pd.read_sql_query(query, connection)
df3.sample(1)

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_1184\3410962354.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df3 = pd.read_sql_query(query, connection)


count
0    294

Количество сфер деятельности в базе 294

***

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [34]:
query = f'''
            select 
            a.name as area_name, count(v.name) as count_vac
            from areas a, vacancies v
            where a.id=v.area_id
            group by 1
            order by 2 desc
            limit 5        
'''
df5 = pd.read_sql_query(query, connection)
df5


C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_10652\1501038170.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df5 = pd.read_sql_query(query, connection)


area_name  count_vac
0           Москва       5333
1  Санкт-Петербург       2851
2            Минск       2112
3      Новосибирск       2006
4           Алматы       1892

Пятерка лидеров: Москва, Санкт_петербург, Минск, Новосибирск, Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [42]:
query = f'''
            select count (name) as cnt_vac
            
            from vacancies
            where salary_from is not Null or salary_to is not NULL           
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_20260\3623160774.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


cnt_vac
0    24073

24073 вакансии имеют хотя бы одно заполненное поле с зарплатой.

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [47]:
query = f'''
            select floor(avg(salary_from)) as avg_salary_f, floor(avg(salary_to)) as avg_salary_t
            
            from vacancies
                      
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_20260\2259317083.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_salary_f  avg_salary_t
0       71064.0      110536.0

Среднее значение зарплаты от 71 064 до 110 536 рублей

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [50]:
query = f'''
            select schedule, employment, count(name)
            
            from vacancies
            group by 1,2
            order by 3 desc            
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_20260\2543875407.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [55]:
query = f'''
            select experience as "Требуемый опыт работы", count(name) as "количество вакансий"
            
            from vacancies
            group by 1
            order by 2          
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_20260\1291560958.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


Требуемый опыт работы  количество вакансий
0           Более 6 лет                 1337
1             Нет опыта                 7197
2         От 3 до 6 лет                14511
3    От 1 года до 3 лет                26152

Количество вакансий в порядке возрастания: Более 6 лет — Нет опыта — От 3 до 6 лет — От 1 года до 3 лет

***

In [ ]:
# выводы по детальному анализу вакансий

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [113]:
query = f'''
            select e.name, count(v.name) as cnt_vacancies
            
            from vacancies v , employers e
            where v.employer_id=e.id
            group by 1
            order by 2 desc
            limit 5
            
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_10652\269498829.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name  cnt_vacancies
0         Яндекс           1933
1     Ростелеком            491
2       Тинькофф            444
3           СБЕР            428
4  Газпром нефть            331

Работадатели: на 1 месте - Яндекс,на 5м - Газпром нефть

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [14]:
query = f'''
            select 
            distinct a.name area_name, count(distinct e.name) employers_cnt, count(distinct v.id) vacancies_cnt
            
            from areas a
            left join VACANCIES v on a.id = v.area_id
            left join employers e on a.id = e.area
            where v.id is null
            group by 1
            order by 2 desc
            limit 1
                      
          
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_13872\3604565758.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


area_name  employers_cnt  vacancies_cnt
0    Россия            410              0

Наибольшее количество работадателей, среди регионов в которых нет вакансий, это  Россия 410 работадателей.

In [171]:
query = f'''
            select *
            from vacancies v       
            limit 5      
'''
df5 = pd.read_sql_query(query, connection)
df5


C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_4288\1541937829.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df5 = pd.read_sql_query(query, connection)


id                                               name  \
0  55312386                                Компьютерный Мастер   
1  55843014                            Системный администратор   
2  54525964                      Lead Java Developer to Poland   
3  54525965                      Lead Java Developer to Poland   
4  55354053  Специалист службы поддержки с техническими зна...   

                                          key_skills          schedule  \
0  Пользователь ПК\tРабота в команде\tРемонт ноут...       Полный день   
1  Средства криптографической защиты информации\t...       Полный день   
2    Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
3    Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
4                                               None  Удаленная работа   

      experience           employment  salary_from salary_to  area_id  \
0      Нет опыта     Полная занятость      64000.0      None     1516   
1  От 3 до 6 лет     Полная занятость          NaN      None      130   
2  От 3 до 6 лет     Полная занятость          NaN      None      160   
3  От 3 до 6 лет     Полная занятость          NaN      None      159   
4      Нет опыта  Частичная занятость      15000.0      None     1955   

   employer_id  
0      5724811  
1      4903713  
2        69961  
3        69961  
4         1740

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [6]:
query = f'''
            select 
            distinct e.name as employer, count(distinct v.area_id) area_name
            from employers e    
            left join VACANCIES v on  e.id=v.employer_id     
                       
            group by 1
            order by 2 desc
            limit 5
            '''
df = pd.read_sql_query(query, connection)
df

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_11592\2704129686.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


employer  area_name
0                  Яндекс        181
1              Ростелеком        152
2              Спецремонт        116
3  Поляков Денис Иванович         88
4               ООО ЕФИН          71

Наибольшее количество регионов у компании Яндекс, 181.

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [90]:
query = f'''
            select 
            count(e.name) as employer, count(distinct ei.industry_id) cnt_industry
            from employers e    
            left join EMPLOYERS_INDUSTRIES ei on e.id=ei.employer_id    
            where ei.industry_id is NULL
                
            order by 2
          
                      
          
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_4288\4236435435.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


employer  cnt_industry
0      8419             0

Сфера деятельности не указана у 8 419 работадателей

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [74]:
query = f'''
            select 
            distinct e.name as employer, count(distinct ei.industry_id) cnt_industry
            from employers e    
            left join EMPLOYERS_INDUSTRIES ei on e.id=ei.employer_id        
            group by 1
            having count(distinct ei.industry_id)=4
            order by 1 
            limit 3          
       
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_4288\3238165656.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


employer  cnt_industry
0  101 Интернет             4
1      21vek.by             4
2          2ГИС             4

На третьем месте находится компания 2ГИС	

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [122]:
query = f'''
            select 
            i.name as industry, count (e.name) as employer
            
            from EMPLOYERS_INDUSTRIES ei    
            left join EMPLOYERS e on e.id=ei.employer_id 
            left join INDUSTRIES i on i.id=ei.industry_id
            group by 1
            having i.name = 'Разработка программного обеспечения'   
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_4288\3859202926.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


industry  employer
0  Разработка программного обеспечения      3553

Количество работадателей, у которых указана сфера деятельности 'Разработка программного обеспечения' - 3553.

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [71]:
query = f'''
        (select 
        a.name as "Регион-миллионник", count(v.id) as vacancies_cnt
        from areas a
        join VACANCIES v on a.id = v.area_id
        where (v.employer_id=1740) and (a.name ='Москва' or a.name ='Санкт-Петербург' or a.name ='Новосибирск' or a.name ='Екатеринбург'or 
        a.name ='Казань'or a.name ='Нижний Новгород'or a.name ='Воронеж'or a.name ='Челябинск'or a.name ='Красноярск'or a.name ='Самара'
        or a.name ='Уфа'or a.name ='Ростов-на-Дону'or a.name ='Омск'or a.name ='Краснодар'or a.name ='Пермь'or a.name ='Волгоград')
        group by 1)
        UNION ALL 
        SELECT
        'total', Count(v.id) 
        from 
        VACANCIES v 
        join areas a on a.id = v.area_id
        where (v.employer_id=1740) and (a.name ='Москва' or a.name ='Санкт-Петербург' or a.name ='Новосибирск' or a.name ='Екатеринбург'or 
        a.name ='Казань'or a.name ='Нижний Новгород'or a.name ='Воронеж'or a.name ='Челябинск'or a.name ='Красноярск'or a.name ='Самара'
        or a.name ='Уфа'or a.name ='Ростов-на-Дону'or a.name ='Омск'or a.name ='Краснодар'or a.name ='Пермь'or a.name ='Волгоград')
        order by 2                                
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_12028\3483929044.py:21: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


Регион-миллионник  vacancies_cnt
0               Омск             21
1          Челябинск             22
2         Красноярск             23
3          Волгоград             24
4              Пермь             25
5             Казань             25
6     Ростов-на-Дону             25
7                Уфа             26
8             Самара             26
9          Краснодар             30
10           Воронеж             32
11       Новосибирск             35
12   Нижний Новгород             36
13      Екатеринбург             39
14   Санкт-Петербург             42
15            Москва             54
16             total            485

У компании Яндекс 485 вакансий в городах миллионниках (16 регионов) Выборка сделана через ID_employer, у компании Яндекс он равен 1740.

In [26]:
query = f'''
            select 
            e.name, v.employer_id, count(v.id)
            from VACANCIES v
            join employers e on e.id=v.employer_id
            group by 1,2
            order by 3 desc
                                       
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_12028\4275622242.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name  employer_id  count
0               Яндекс         1740   1933
1           Ростелеком         2748    491
2             Тинькофф        78638    444
3                 СБЕР         3529    428
4        Газпром нефть        39305    331
...                ...          ...    ...
14901         Прогресс      3122660      1
14902         Прогресс      3202878      1
14903         Прогресс      3998198      1
14904    ПРОГРЕСС АГРО      4446516      1
14905  Прогресс-Гарант      3228623      1

[14906 rows x 3 columns]

In [ ]:
# результат запроса

***

In [ ]:
# выводы по анализу работодателей

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [4]:
query = f'''
            select 
            count(v.id) as vacancies
            
            from VACANCIES v   
            where (name like '%%Данн%%' or name like '%%данн%%' or name like '%%data%%' or name like '%%Data%%' or name like '%%DATA%%' or name like '%%ДАНН%%')
            
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_4572\2390402055.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


vacancies
0       1771

Количество вакансий datascince 1771.

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [5]:
query = f'''
    select
	count(v.name)
    from vacancies v

     where ((lower(v.name) like '%data scientist%' or lower(v.name) like '%data science%' or lower(v.name) like '%исследователь данных%' or
    (v.name like '%ML%' and 
    upper(v.name) not like '%HTML%') or lower(v.name) like '%machine learning%' or lower(v.name) like '%машинн%обучен%')) and 
    (v.name ~* 'junior' or v.experience = 'Нет опыта' or v.employment = 'Стажировка')
            
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_4572\68774771.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0     51

Количество вакансий 51, т.е  для начинающих порог для входа очень сильно ограничен, это всего лишь 2,88% от обего количества вакансий.

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [6]:
query = f'''
        select
        count(*)
        from VACANCIES v
        where ((lower(v.name) like '%data scientist%' or lower(v.name) like '%data science%' or lower(v.name) like '%исследователь данных%' or
        (v.name like '%ML%' and 
        upper(v.name) not like '%HTML%') or lower(v.name) like '%machine learning%' or lower(v.name) like '%машинн%обучен%')) 
        and (v.key_skills ilike '%postgres%' or v.key_skills ilike '%SQL%')
            
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_4572\948650469.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0    201

Количество вакансий 201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [8]:
query = f'''
            select 
            count(v.name) as vacancies
            
        from VACANCIES v   
        where ((lower(v.name) like '%data scientist%' or lower(v.name) like '%data science%' or lower(v.name) like '%исследователь данных%' or
        (v.name like '%ML%' and upper(v.name) not like '%HTML%') or lower(v.name) like '%machine learning%' or lower(v.name) like '%машинн%обучен%'))
        and key_skills like '%Python%'
            
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_4572\4121460684.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


vacancies
0        351

Количество вакансий 351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [147]:
query = f'''
        select 
       round(AVG((length((replace(key_skills,CHR(9),'**')))-length(key_skills))+1),2) as avg_key_skills
        
      
        from VACANCIES v   
        where ((lower(v.name) like '%data scientist%' or lower(v.name) like '%data science%' or lower(v.name) like '%исследователь данных%' or
        (v.name like '%ML%' and upper(v.name) not like '%HTML%') or lower(v.name) like '%machine learning%' or lower(v.name) like '%машинн%обучен%'))
        
            
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_4572\4251846734.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_key_skills
0            6.41

В среднем 6,41 ключевых навыков.

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [152]:
query = f'''
        select 
        distinct experience, 
        round(AVG(coalesce((salary_from+salary_to)/2, salary_from, salary_to)),0)  as AVG_salary
        
    
        from VACANCIES v   
        where ((lower(v.name) like '%data scientist%' or lower(v.name) like '%data science%' or lower(v.name) like '%исследователь данных%' or
        (v.name like '%ML%' and upper(v.name) not like '%HTML%') or lower(v.name) like '%machine learning%' or lower(v.name) like '%машинн%обучен%'))
        and (salary_from is not null or salary_to is not null)
        group by 1   
'''
df = pd.read_sql_query(query, connection)
df

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_4572\3802420638.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


experience  avg_salary
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0
2       От 3 до 6 лет    243115.0

Средняя зарплата 243115 руб.

***

# Общий вывод по проекту

In [165]:
query = f'''
        select experience, count(id) as cnt_vacancies, round(AVG(coalesce((salary_from+salary_to)/2, salary_from, salary_to)),0)  as AVG_salary,
        round(AVG((length((replace(key_skills,CHR(9),'**')))-length(key_skills))+1),2) as avg_key_skills , count(distinct employer_id) as cnt_employers    
        from VACANCIES v   
        where ((lower(v.name) like '%data scientist%' or lower(v.name) like '%data science%' or lower(v.name) like '%исследователь данных%' or
        (v.name like '%ML%' and upper(v.name) not like '%HTML%') or lower(v.name) like '%machine learning%' or lower(v.name) like '%машинн%обучен%'))
        group by 1
            
'''
df = pd.read_sql_query(query, connection)
df
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

C:\Users\e.korenkova\AppData\Local\Temp\ipykernel_4572\3957439571.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


experience  cnt_vacancies  avg_salary  avg_key_skills  \
0         Более 6 лет             34         NaN            5.94   
1           Нет опыта             28     74643.0            7.42   
2  От 1 года до 3 лет            223    139675.0            6.73   
3       От 3 до 6 лет            195    243115.0            5.99   

   cnt_employers  
0              9  
1             23  
2            132  
3             99

    В своих исследованиях я провела анализ вакансий по специалистам datascince, вот следующие выводы:
1. Порог для входа очень высокий (самые высокие требования по количеству навыков и самое маленькое количество вакансий и работодателей на рынке);
2. Больше всего восстребованы специалисты с опытом работы от 1 до 3х лет;
3. Чем выше требования к опыту, тем меньшее количество навыков присутствует, скорее всего это узкая специализация специалистов;
4. Есть прямая корреляция уровня заработной платы от опыта. Чем больше опыт, тем выше заработная плата. Отсутствие данных по зарплате свыше 6 лет, говорит о том, что зарплата определяется по договоренности с кандидатом.



In [72]:
connection.close()